### ISSUE TITLE
Feature request: Negative binomial with mean and dispersion parameterization

### ISSUE THREAD

**Issue by gokceneraslan** [tensorflow/probability#372]
For many people working with count data (especially in ecology and genomics), it is very common to parameterize negative binomial with mean (`total_count*exp(logits)`) and dispersion (`1/total_count`). Would it be possible to add this reparametrization of NB to tfp?

**Comment by brianwa84** [comment]
It seems straightforward to do the reparameterization since one can infer
the existing parameters from those you identify. Are there any numerical
stability benefits or downsides to this parameterization? e.g. It seems
like you might lose precision on total_count if it were represented by it's
inverse between 0 and 1.

https://github.com/tensorflow/probability/blob/c6c14ff4e684ebc2613833d71cb476ba5c27e2a8/tensorflow_probability/python/distributions/negative_binomial.py#L148
is where we do sampling.

**Comment by gokceneraslan** [comment]
There is no consensus over how dispersion is defined actually, some people use `number of failures` directly, some people use the inverse. Former approaches `+inf` in Poisson regime, so I'm not entirely sure if this is as bad as it approaching zero.

R uses both parameterization in following way: `rnbinom(n, size, prob, mu)` where size is the target for number of successful trials, or dispersion parameter (the shape parameter of the gamma mixing distribution) and mu is the mean. Either prob or mu has to be specified by the user, but not both. Details are [here](https://stat.ethz.ch/R-manual/R-patched/library/stats/html/NegBinomial.html).

I think having both parameterizations would be very convenient in tfp too.

**Comment by brianwa84** [comment]
I agree, if it's a fairly common parameterization, we should support it. It seems like R does support this, but scipy does not. 

If you'd like to send a PR to add either alternative constructor args, I think we'd be open to it (make sure to raise if >1 parameterization is actually used). It would be good if you could could put a reference or two [in the docstring](https://github.com/tensorflow/probability/blob/master/STYLE_GUIDE.md#tensorflow-probability-style) showing how the new parameterization is used in practice in a couple compbio/eco papers? I found this one: https://stackoverflow.com/questions/40846992/alternative-parametrization-of-the-negative-binomial-in-scipy

We try to be sparing with new arg names, but dispersion seems fairly commonly used for this and is somewhat analagous to `concentration`, which we use widely. @jvdillon thoughts?

**Comment by jvdillon** [comment]
We've discussed mean parameterization of distributions. Unfortunately there's a few usability questions which need discussion. Can we discuss this as a team and report back?

**Comment by gokceneraslan** [comment]
Sure. I think dispersion is the easy one, given that one can use `total_count` drirectly as dispersion. But `mean` is tricker.

**Comment by gokceneraslan** [comment]
Here is Stan's alternative NB parameterization: https://mc-stan.org/docs/2_19/functions-reference/nbalt.html

**Comment by jvdillon** [comment]
Hi @gokceneraslan.

Our team has discussed the design possibilities and come up with these five options:

**Comment by jvdillon** [comment]
It sounds like our team is converging on favoring #3. We'd still like to hear your opinion though. :)

**Comment by gokceneraslan** [comment]
Looks good. A few questions:

- How would that look for negative binomial?
- What about the other parameters like scale in location-scale family distiburions? 
- How about `tfd.Normal.from_location_scale(location=0.25, scale=1)`?

**Comment by brianwa84** [comment]
For NB I think it would look like:
`tfd.NegativeBinomial.from_mean_dispersion(mean=xx, dispersion=yy)`

**Comment by brianwa84** [comment]
Such function would need a nice docstring to explain precisely which dispersion parameter it's talking about. PR welcome.

```
@classmethod
def from_mean_dispersion(cls, mean, dispersion, **kwargs):
  ...
  if 'logits' in kwargs or 'total_count' in kwargs: raise
  kwargs['total_count']=...
  kwargs['logits']=...
  ...
  return cls(**kwargs)
```

**Comment by jvdillon** [comment]
Brian, would you be ok with just "from_mean"?  Or perhaps: "mean_parameterization"? I ask because it'd be nice if we could use names which were the same across all distributions.  I also dont think there's diminishing value in repeating the kwargs as part of the name.  (Saying "mean" is different because it conveys the semantics.)

Additionally, I think its ok that the mean parametrization takes other args. For example, a normal distribution takes a scale but we pin the scale to a constant in generalized linear regression. (Making it mean parameterized.)

Thoughts?

**Comment by brianwa84** [comment]
I'm ok with from_mean as long as it is unambiguously *the* mean
parameterization.

**Comment by jarotter** [comment]
Did this parametrization ever get to tfp?

**Comment by alexonov** [comment]
Is there any update ion this?

**Comment by brianwa84** [comment]
No updates. Still nice to have, PRs would be welcome. Consider using `tfp.util.DeferredTensor` to defer any transformations to happen inside a `GradientTape`.
i.e.
```
nb = NegativeBinomial.from_mean_dispersion(mu, disp)
with tf.GradientTape() as t:
  t.watch((mu, disp))
  lp = nb.log_prob(3)
t.gradient(lp, (mu, disp))  # not None
```

**Comment by jvdillon** [comment]
As another example, I think you could more or less copy this:

https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/glm/family.py#L512
